### PR TITLE
FIR checker: report incorrect uses of type parameter in callable references

### PR DESCRIPTION
**Pull request by jsjeon**
The main motivation is to report `CALLABLE_REFERENCE_LHS_NOT_A_CLASS`, e.g., `T::toString`, after `CLASS_LITERAL_LHS_NOT_A_CLASS` (e.g., `A<T>::class`) was added at #4162. Unlike class literals, which are desugared to `FirGetClassCall`, callable reference is a kind of qualified expression, and LHS is represented as an explicit receiver of qualified expression. That explicit receiver can be either resolved qualifier, resolved reified type parameter, or another qualified expression (and the last part caused issues that I'll explain later).

For the case like `T::toString`, `T` is a resolved qualifier without a return type. It's already regarded as an error in that case, but with a default-ish `Other` kind. To distinguish an incorrect use of type parameter in a qualified expression, I introduced `ConeTypeParameterInQualifiedAccess`. That differentiation, though, revealed more diagnostic candidates than expected, e.g., `T.foo` or just `T` as an expression. As a result, the 1st commit covers:
* CALLABLE_REFERENCE_LHS_NOT_A_CLASS
* TYPE_PARAMETER_ON_LHS_OF_DOT
* TYPE_PARAMETER_IS_NOT_AN_EXPRESSION

For that last one, we should make sure `T` we're visiting now is not a part of another qualified expression. Otherwise, e.g., for `T.foo`, we would report last two diagnostics.

------

Then, two changes caught my attention: `T?::class` and `T::class`, and found three issues. First, the former should be reported as `NULLABLE_TYPE_IN_CLASS_LITERAL_LHS`, but my initial work (#4162) focused too much on cases like `A?::class`. Plus, I realized something like `exp::class` (where `exp` is nullable) should be handled too. The 2nd commit generalizes a diagnostic for nullable LHS of class literals, and introduces:
* EXPRESSION_OF_NULLABLE_TYPE_IN_CLASS_LITERAL_LHS

------

Second, the latter (`T::class`) needs a different diagnostic if it's not reified. The 3rd commit introduces:
* TYPE_PARAMETER_AS_REIFIED

------

Third, both cases still report `TYPE_PARAMETER_IS_NOT_AN_EXPRESSION` on `T`, which shouldn't be the case. The root cause is, class literals, which is a special kind of callable reference (a subtype of qualified access), are not properly tracked as qualified access. Note that `FirGetClassCall` spans Java use (e.g., `X.class`), so organizing hierarchy again (like making `FirGetClassCall` as a subtype of `FirCallableReferenceAccess`) would be infeasible or non-trivial. Thus, the 4th commit adds tracking of `FirGetClassCall` in checker context, and use it to filter out `T` in class literals too.

------

Finally, there was a wrong change like `fun <T> test1() = T::class`, which needs a different diagnostic. The 5th commit introduces:
* KCLASS_WITH_NULLABLE_TYPE_PARAMETER_IN_SIGNATURE

### CODE REVIEW

File: idea/idea-frontend-fir/src/org/jetbrains/kotlin/idea/frontend/api/fir/diagnostics/KtFirDataClassConverters.kt
Comment (jsjeon) [inline idea/idea-frontend-fir/src/org/jetbrains/kotlin/idea/frontend/api/fir/diagnostics/KtFirDataClassConverters.kt:1018]: 🤦  Ah, hope it will be added by someone else soon!

File: compiler/testData/diagnostics/tests/declarationChecks/kClassInSignature.fir.kt
Comment (jsjeon) [inline compiler/testData/diagnostics/tests/declarationChecks/kClassInSignature.fir.kt:11]: So, this one is the only regression for now. It should be `UPPER_BOUND_VIOLATED`, but I'd like to make a check point here, and will see related diagnostics together.

Comment (jsjeon) [reply compiler/testData/diagnostics/tests/declarationChecks/kClassInSignature.fir.kt:11]: Looks like we already have an expression-level checker: https://github.com/JetBrains/kotlin/blob/master/compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirUpperBoundViolatedChecker.kt but the same idea should be applied to the return type of declarations. May need more study/refactoring.

File: idea/idea-frontend-fir/src/org/jetbrains/kotlin/idea/frontend/api/fir/diagnostics/KtFirDataClassConverters.kt
Comment (jsjeon) [reply idea/idea-frontend-fir/src/org/jetbrains/kotlin/idea/frontend/api/fir/diagnostics/KtFirDataClassConverters.kt]: Added here: https://github.com/JetBrains/kotlin/commit/1d42c7005696df6188c28d689c28df9b4dcb5604

File: compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/transformers/FirCallCompletionResultsWriterTransformer.kt
Comment (tgeng) [inline compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/transformers/FirCallCompletionResultsWriterTransformer.kt:97]: It looks like reified type parameters are not reported. For example

```
inline fun <reified T> foo() {
    T::toString // error
}
```

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirTypeParameterInQualifiedAccessChecker.kt
Comment (tgeng) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirTypeParameterInQualifiedAccessChecker.kt:39]: Looks like this is because `org.jetbrains.kotlin.fir.analysis.collectors.AbstractDiagnosticCollector` does not override `visitErrorTypeRef`, which, by default, visits resolved type ref and hence skips reporting errors.

It looks like a bug to me and should be fixed with an overridden visitErrorTypeRef. For example

```
override fun visitErrorTypeRef(errorTypeRef: FirErrorTypeRef, data: Nothing?) {
    errorTypeRef.delegatedTypeRef?.accept(this, data)
}
```

File: compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/transformers/FirCallCompletionResultsWriterTransformer.kt
Comment (jsjeon) [reply compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/transformers/FirCallCompletionResultsWriterTransformer.kt]: Thanks, updated `compiler/testData/diagnostics/tests/generics/tpAsReified/InFunction.kt` as it contains similar (yet more undetected) cases.

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirTypeParameterInQualifiedAccessChecker.kt
Comment (jsjeon) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirTypeParameterInQualifiedAccessChecker.kt:40]: Thanks for the question. I think my answers are, everything is intentional.

> Should this check be done in org.jetbrains.kotlin.fir.analysis.diagnostics.ConeDiagnosticToFirDiagnosticKt#toFirDiagnostic instead?

It's the place a cone diagnostic can be mapped to a FIR diagnostic without any context, while this one needs to be reported differently, depending on the context, like whether the containing expression is a callable reference, whether this type parameter is a standalone expression, etc.

Re: `AbstractDiagnosticCollector#visitErrorTypeRef`

I'm afraid it will generate lots of redundant/repeated diagnostics.

The reason this diagnostic is not visited is, we intentionally change its source as _fake_ source (as previously mentioned in the code comment).

Comment (tgeng) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirTypeParameterInQualifiedAccessChecker.kt:40]: I see. Thanks for the clarification!

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirGetClassCallChecker.kt
Comment (jsjeon) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirGetClassCallChecker.kt]: Yeah, don't know why I added that in the beginning. Removed at the following commit.

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirGetClassCallChecker.kt
Comment (jsjeon) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirGetClassCallChecker.kt:45]: Thanks. I noticed similar examples, but didn't remember why I assumed it would require more learning or something. Anyway, yeah, unless `T` is reified or has a subtype constraint with non-null type, we should report nullable expression of LHS of class literals. Alas(?), checking a subtype of `Any` is added as part of `KClass` signature, so it's a bit hard to figure out commits' order and/or merging. Hence, added as the last commit.

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirKClassWithIncorrectTypeArgumentChecker.kt
Comment (jsjeon) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirKClassWithIncorrectTypeArgumentChecker.kt:39]: Good catch! Actually, member function is also not reported in FE 1.0. It turns out that callable member descriptor is somewhat stronger than I thought. It literally means this checker should be applied to top-level declarations only. Updated tests and checker accordingly.

**Review by tgeng — APPROVED** [review]
LGTM. But please also wait for Mark's approval.

**Comment by jsjeon** [comment]
> LGTM. But please also wait for Mark's approval.

Sure, cc @punzki

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirGetClassCallChecker.kt
Comment (punzki) [inline compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirGetClassCallChecker.kt:82]: What about the following case:
`fun <T> f2(t: T): Any where T : Any?, T : Comparable<T> = t::class`

This is ok in FE1.0 but would report the error with this change, since `Any?` is not a subtype of `Any`. I think this should be either `bounds.all` or get the intersection type.

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirGetClassCallChecker.kt
Comment (punzki) [inline compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirGetClassCallChecker.kt:38]: Maybe you can update the examples with type parameter examples (reified and not reified)?

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirGetClassCallChecker.kt
Comment (punzki) [inline compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirGetClassCallChecker.kt:92]: Nit: Just my opinion but the name is a little confusing, I think just `asTypeParameterSymbol` is ok.

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirKClassWithIncorrectTypeArgumentChecker.kt
Comment (punzki) [inline compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirKClassWithIncorrectTypeArgumentChecker.kt:36]: Can a top-level declaration be local?

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirKClassWithIncorrectTypeArgumentChecker.kt
Comment (punzki) [inline compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirKClassWithIncorrectTypeArgumentChecker.kt:39]: Is this still needed? Don't think you can have an accessor as top-level in a file.

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirKClassWithIncorrectTypeArgumentChecker.kt
Comment (punzki) [inline compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirKClassWithIncorrectTypeArgumentChecker.kt:31]: Some examples in comments would be useful.

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirKClassWithIncorrectTypeArgumentChecker.kt
Comment (punzki) [inline compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirKClassWithIncorrectTypeArgumentChecker.kt:62]: Same as comment on other commit: What about the following case:
`fun <T> f3() where T : Any?, T : Comparable<T> = T::class`

I think this should be either `bounds.all` or get the intersection type.

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirGetClassCallChecker.kt
Comment (jsjeon) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirGetClassCallChecker.kt]: I forgot such upper bounds are intersection, thanks! We can simply rely on subtype check, simple, clearer! Also, updated a test.

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirGetClassCallChecker.kt
Comment (jsjeon) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirGetClassCallChecker.kt:39]: These are about unparsed `?`. Rather, I added some examples somewhere else when new diagnostics are added.

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirKClassWithIncorrectTypeArgumentChecker.kt
Comment (jsjeon) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirKClassWithIncorrectTypeArgumentChecker.kt]: Nope. This should be removed together when I switch to check top-level declarations. Thanks.

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirKClassWithIncorrectTypeArgumentChecker.kt
Comment (jsjeon) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirKClassWithIncorrectTypeArgumentChecker.kt:40]: Yes, needed: https://github.com/JetBrains/kotlin/blob/master/compiler/testData/diagnostics/tests/declarationChecks/kClassInSignature.kt#L7

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirKClassWithIncorrectTypeArgumentChecker.kt
Comment (jsjeon) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirKClassWithIncorrectTypeArgumentChecker.kt]: Right. Again, simple subtype check is good enough (and match FE 1.0). Added a test as well, thanks!

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirGetClassCallChecker.kt
Comment (jsjeon) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirGetClassCallChecker.kt:96]: I think `safe` emphasizes the nullable return type. E.g.,

* https://github.com/JetBrains/kotlin/blob/master/compiler/fir/tree/src/org/jetbrains/kotlin/fir/expressions/FirOperation.kt#L36
* https://github.com/JetBrains/kotlin/blob/master/core/util.runtime/src/org/jetbrains/kotlin/utils/addToStdlib.kt#L73
* https://github.com/JetBrains/kotlin/blob/master/compiler/fir/tree/src/org/jetbrains/kotlin/fir/types/FirTypeUtils.kt#L20

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirGetClassCallChecker.kt
Comment (jsjeon) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirGetClassCallChecker.kt:42]: `?` can be missed during the desugaring, and that field is updated during the resolution. So there are chances that those are not matched. Also answered here: https://github.com/JetBrains/kotlin/pull/4162#discussion_r585106944

**Review by mglukhikh — APPROVED** [review]
Generally it looks good. I don't like some minor things, for example, using `isSubtypeOfAny` effectively to check that type is not null looks awkward a bit. I will investigate and fix these details myself.

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirKClassWithIncorrectTypeArgumentChecker.kt
Comment (punzki) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirKClassWithIncorrectTypeArgumentChecker.kt:40]: In that case, isn't it the `FirProperty` that is top-level? Not the accessor.

**Comment by mglukhikh** [comment]
Merged manually with some minor changes:
- https://github.com/JetBrains/kotlin/commit/0b9598d880437b4ca8eb2d07b905c31eab67e6eb
- https://github.com/JetBrains/kotlin/commit/599a0d9c2b2a9cc9d3df49076e9bd2e8e9ebf464
- https://github.com/JetBrains/kotlin/commit/0e8f8362da34e212c7a728728e126f0a4b9c7945

**Pull request closed without merging**
